### ISSUE TITLE
Should offer mongodb-10gen packages

### ISSUE THREAD

**Issue by yfeldblum** [edelight/chef-mongodb#1]
10gen [publishes](http://www.mongodb.org/display/DOCS/Ubuntu+and+Debian+packages) debian/ubuntu packages for mongodb, which are versions ahead of the packages in the debian/ubuntu repositories.

The mongodb cookbook should offer an option to add the 10gen apt repository and install the 10gen package.

**Comment by tobami** [comment]
Hi,

we use ourselves the 10-gen packages, but it is a debian specific solution, so we didn't include it. You can just add your own recipe to the run list that then adds the repo. The mongodb cookbook will then pull the newer packages.

**Comment by yfeldblum** [comment]
To clarify, is your suggestion that I do something like the following in a site-cookbook:

```
apt_repository "10gen" do
  # ...
end

package "mongodb" do
  package_name "mongodb-10gen"
end

include_recipe "mongodb"
```

**Comment by skippy** [comment]
hey,
this is an 'opinionated' comment, but what about hard-coding the version # and then pulling the specific version down as a .tgz.

The thing that makes me wary about getting something like this from apt-get is that you cannot specify the version; so if you add an instance to a set and 10-gen has up'ed the stable version, say from 1.8.2 to 2.0.0, you could be hosed.  For something this interconnected, I think specifying the exact version is a safer bet.

if you agree, I'll submit a patch.
thanks!

**Comment by yfeldblum** [comment]
The version number should not be hardcoded. The best that may be done is to add a cookbook attribute with a default value for the mongoid version.

However, that's not strictly necessary either: you can do that in your cookbook and override what's done in the mongodb cookbook.

```
include_recipe "mongodb"

# During the compile phase, this reaches into the `package "mongodb"`
#   resource as created in the `mongodb::default` cookbook and updates
#   or overrides the attributes.
# When the resource is actually executed, during the execute phase, the
#   package resource will be executed with the attributes we set here.
resources(:package => "mongodb").instance_exec do
  version "2.0.0"
end
```

**Comment by skippy** [comment]
huh!  I didn't know that you could do that with the override; nice trick!

but that doesn't work for apt-get; it doesn't store previous versions.  If I modify the default.rb file directly with:

```
package "mongodb" do
  action :install
  version "1.6.0"
end
```

it still installs 2.0.0.  So if I was to say, launch another node into a replica set, last week they would have been 1.8.x and now they would be 2.0.0.  

not a huge deal ;easy to fix in a fork (and why I started with 'opinionated' ;), but thought I would bring it up.  Let me know f you change your mind and I'll submit a branch

**Comment by tobami** [comment]
@yfeldblum: I just added a new 10gen_repo recipe. It is only implemented for Ubuntu/Debian, but it should be easy to expand support for other distros.

@skippy: The version issue you mention is legitimate. I do think, however, that 10gen has a wrong approach there. The way Opscode does it now is much better. If you are not going to use the official distribution packages, you want to be sure that users of the repository won't have nasty surprises, and you can only ensure that with a different repo pro major version.

I am closing the issue despite missing 10gen support for other distros. Feel free to open a new one for that, or just sending a pull request.
Thanks for the feedback!

Closed by 343b2a5
